# scRNAseq-practical-course

This repository holds a portion of the materials for the Practical Course: for the graduate students and postdocs of


Practical Course: Single-Cell RNA Sequencing Data Analysis
13.01.2022 - 14.01.2022, 09:00 Uhr - 17:00 Uhr

![Course syllabus](https://htmlpreview.github.io/?https://github.com/danieljgorski/scRNAseq-practical-course/blob/main/Course-syllabus.html)
Link IRTG
Link SFB

The aim of this course is to introduce single-cell RNA sequencing (scRNAseq) data analysis to individuals with little or no experience with scRNAseq or R programming. Our intent is to provide foundational knowledge that can be expanded to other computational methods or simply help individuals better critically evaluate scRNAseq studies. We will provide a short overview of scRNAseq technologies, experimental design and typical workflows. Using Seurat, a popular single-cell genomics toolkit, we will guide students through the basic analysis steps, starting from the expression count matrix to cluster annotation.



This repository covers lecture-2, lab-2 focused data analysis


Guided workflow that relies heavily on Seurat vingetttes (link Seurat) with extra tips

They cover:
*cluster etc...


Instructions

Download Lab-2 data here,
Clone repository
Place data inside Lab-2

#Outputdir results in Lab-2.R code
Create a results f inside Lab-2
